aten/src/ATen/native/TensorTransformations.cpp

```diff
@@ -9,51 +10,90 @@
 namespace at {
 namespace native {
 
+template <typename scalar_t>
+void inline flip_cpu_kernel(
+  const int64_t numel,
+  const int64_t total_dims,
+  const int64_t flip_dims_size,
+  const int64_t* stride_contiguous_d,
+  const int64_t* flip_dims_d,
+  const int64_t* strides_d,
+  const int64_t* sizes_d,
+  const scalar_t* in_tensor_d,
+  scalar_t* out_tensor_d
+){
+  int64_t i;
+
+  #pragma omp parallel for private(i) if (numel > 1000)
+  for (i = 0; i < numel; i++) {
+    int64_t cur_indices = i;
+    int64_t rem = 0;
+    int64_t dst_offset = 0;
+
+    for (int64_t d = 0; d < total_dims; d++) {
+      int64_t temp = cur_indices;
+      cur_indices = cur_indices / stride_contiguous_d[d];
+      rem = temp - cur_indices * stride_contiguous_d[d];
+      for (int64_t j = 0; j < flip_dims_size; j++) {
+        if (d == flip_dims_d[j]) cur_indices = sizes_d[d] - 1 - cur_indices;
+      }
+      dst_offset += cur_indices * strides_d[d];
+      cur_indices = rem;
+    }
+    out_tensor_d[i] = in_tensor_d[dst_offset];
+  }
+}
+
 Tensor flip_cpu(const Tensor& self, IntList dims) {
-  const int64_t total_dims = self.dim(), flip_dims_size = dims.size();
-  flip_check_errors(total_dims, flip_dims_size, dims);
+  auto in_tensor = self;
+  const int64_t total_dims = self.dim();
 
+  dim_list_to_bitset(dims, total_dims);
   auto flip_dims_v = dims.vec();
-  wrap_all_dims(flip_dims_v, total_dims);
-  std::sort(flip_dims_v.begin(), flip_dims_v.end());
-  auto final_indices = std::vector<at::Tensor>(total_dims);
-
-  auto indices = std::vector<at::Tensor>(flip_dims_size);
-  for (int64_t i = 0; i < flip_dims_size; i++) {
-    indices[i] = at::arange(self.size(flip_dims_v[i]) - 1, -1, -1, self.type().toScalarType(at::kLong));
-    // creates a meshgrid
-    auto temp = std::vector<int64_t>(flip_dims_size, 1);
-    temp[i] = indices[i].size(0);
-    indices[i] = indices[i].view(IntList(temp));
-    final_indices[flip_dims_v[i]] = indices[i];
-  }
-
-  // check if distance between two flip dims >= 2, where permute of output tensor is needed,
-  // because the advanced indexing puts all non-consecutive indices in the beginning of the tensor
-  bool to_permute = false;
-  int64_t first = flip_dims_v[0], second = flip_dims_v[0];
-  for (int64_t i = 1; i < flip_dims_size; i++) {
-    second = flip_dims_v[i];
-    if (second - first >= 2) {
-      to_permute = true;
-      break;
+  maybe_wrap_dims(flip_dims_v, total_dims);
+
+  const int64_t numel = in_tensor.numel();
+  auto strides = in_tensor.strides();
+  auto strides_v = strides.vec();
+  auto strides_t = at::CPU(kLong).tensorFromBlob(strides_v.data(), {static_cast<int64_t>(strides_v.size())});
```

Comment:
you don't need to make these tensors because this is cpu

Reply:
you're right! Will change it accordingly

Reply:
changed it to pass in Tensor now, but for reasons I don't understand, now the runtime is longer on the same benchmark:
```
>>> a = torch.randn(1000, 1000)
>>> %timeit -r 100 a.flip(0, 1)
22.2 ms ± 2.88 ms per loop (mean ± std. dev. of 100 runs, 100 loops each)
```

Did I do something wrong?